### PR TITLE
feat: registry proxy

### PR DESCRIPTION
**Pull request by DmitriyMV**
Implement container registry proxy.

### CODE REVIEW

File: hack/cri-plugin.part [hunk 1]
```diff
@@ -1,4 +1,2 @@
 version = 3
 
-[plugins."io.containerd.cri.v1.images"]
-  discard_unpacked_layers = true
```
Comment (smira) [inline hack/cri-plugin.part:4]: we don't need this and matching changing `image.go`

File: internal/app/machined/pkg/system/services/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+	mux.HandleFunc("GET /v2/{args...}", s.serveHTTP)
+
+	giveOk := func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(http.StatusOK) }
+	for _, p := range []string{"v2", "healthz"} {
+		mux.HandleFunc("GET /"+p, giveOk)
+		mux.HandleFunc("GET /"+p+"/{$}", giveOk)
+	}
+
+	server := http.Server{Addr: ":3172", Handler: mux}
```
Comment (smira) [inline internal/app/machined/pkg/system/services/registry/registry.go:51]:
```suggestion
	server := http.Server{Addr: "localhost:3172", Handler: mux}
```

File: internal/app/machined/pkg/system/services/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+// Run is an entrypoint to the API service.
+func (s *Service) Run(ctx context.Context) error {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc("GET /v2/{args...}", s.serveHTTP)
+
+	giveOk := func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(http.StatusOK) }
+	for _, p := range []string{"v2", "healthz"} {
+		mux.HandleFunc("GET /"+p, giveOk)
```
Comment (smira) [inline internal/app/machined/pkg/system/services/registry/registry.go:47]: I'm not sure if `mux` handles `HEAD` transparently, but I believe they do `HEAD /v2` requests as well

File: internal/app/machined/pkg/system/services/registry/registry.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+	return manifest.MediaType, manifestBlob, nil
+}
+
+type canonical struct {
+	reference.NamedTagged
+	digest digest.Digest
+}
+
+func (c *canonical) String() string        { return c.NamedTagged.String() + "@" + c.digest.Encoded() }
+func (c *canonical) Digest() digest.Digest { return c.digest }
```
Comment (dsseng) [inline internal/app/machined/pkg/system/services/registry/registry.go:248]: Weird formatting, won't classic one with newlines look less out of place?

File: internal/app/machined/pkg/system/services/registry/store_darwin.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (dsseng) [inline internal/app/machined/pkg/system/services/registry/store_darwin.go:1]: Does this module get pulled into talosctl build? Then perhaps needs a Windows counterpart, else why does it need a Darwin implementation?

File: internal/app/machined/pkg/system/services/apid.go [hunk 6]
```diff
@@ -53,7 +53,7 @@ func (o *APID) ID(r runtime.Runtime) string {
 }
 
 // apidResourceFilter filters access to COSI state for apid.
-func apidResourceFilter(ctx context.Context, access state.Access) error {
+func apidResourceFilter(_ context.Context, access state.Access) error {
```
[the remark this answers is not in the document]
Comment (DmitriyMV) [reply internal/app/machined/pkg/system/services/apid.go]: Moved to the https://github.com/siderolabs/talos/pull/9760

File: internal/app/machined/pkg/system/services/registry/registry.go (hunk 4 above)
Comment (DmitriyMV) [reply internal/app/machined/pkg/system/services/registry/registry.go:248]: Usually gofmt allows functions with 105 characters or less to be single line.

**Review by dsseng — APPROVED** [review]

File: internal/app/machined/pkg/system/services/registry/app/main.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+func app() error {
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	development, err := zap.NewDevelopment()
+	if err != nil {
+		return fmt.Errorf("failed to create development logger: %w", err)
+	}
+
+	homeDir, err := os.UserHomeDir()
```
Comment (frezbo) [inline internal/app/machined/pkg/system/services/registry/app/main.go:35]: why do we need the homedir here? i guess it's `/root` in talos :thinking:

Comment (DmitriyMV) [reply internal/app/machined/pkg/system/services/registry/app/main.go:35]: It's just an standalone example that you can run and test against.

File: internal/app/machined/pkg/system/services/registryd.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+				logging.WithColoredLevels(),
+			)),
+		).Run(ctx)
+	}, runner.WithLoggingManager(r.Logging())), nil
+}
+
+// HealthFunc implements the HealthcheckedService interface.
+func (m *Registryd) HealthFunc(runtime.Runtime) health.Check {
+	return func(ctx context.Context) error {
+		// TODO: implement me
```
Comment (smira) [inline internal/app/machined/pkg/system/services/registryd.go:72]: we should just do TCP/HTTP check here, should be trivial, e.g. like `kubelet` health check

File: internal/app/machined/pkg/system/services/registry/registry.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/containerd/containerd/v2/core/content"
+	"github.com/distribution/reference"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/siderolabs/gen/xerrors"
+	"go.uber.org/zap"
+)
+
+// NewService creates a new instance of the registry service.
+func NewService(root string, logger *zap.Logger) *Service {
```
Comment (smira) [inline internal/app/machined/pkg/system/services/registry/registry.go:32]: let's add multiple roots (probably best if we can inject them on the fly? or the service itself can read that from the resource, but that can come later)

**Review by smira — APPROVED** [review]

File: internal/app/machined/pkg/system/services/registry/registry.go (hunk 9 above)
Comment (smira) [inline internal/app/machined/pkg/system/services/registry/registry.go:32]:
```suggestion
func NewService(rootsFunc func() (paths []string, err error), logger *zap.Logger) *Service {
```

**Review by smira — APPROVED** [review]

**Review by frezbo — APPROVED** [review]

File: internal/app/machined/pkg/system/services/registry/registry.go (hunk 9 above)
Comment (frezbo) [reply internal/app/machined/pkg/system/services/registry/registry.go]: this probably needs to be added

Comment (smira) [reply internal/app/machined/pkg/system/services/registry/registry.go]: it was added, but via MultiFS iterator (see the example in the service start function)

Comment (frezbo) [reply internal/app/machined/pkg/system/services/registry/registry.go]: oh i thought I commented on the healthcheck part, sorry

File: internal/app/machined/pkg/system/services/registryd.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+// NewRegistryD returns a new docker mirror registry service.
+func NewRegistryD() system.Service                                       { return &registryD{} }
+func (r *registryD) ID(runtime.Runtime) string                           { return "registryd" }
+func (r *registryD) HealthSettings(runtime.Runtime) *health.Settings     { return &health.DefaultSettings }
+func (r *registryD) PreFunc(context.Context, runtime.Runtime) error      { return nil }
+func (r *registryD) PostFunc(runtime.Runtime, events.ServiceState) error { return nil }
+func (r *registryD) Condition(runtime.Runtime) conditions.Condition      { return nil }
+func (r *registryD) DependsOn(runtime.Runtime) []string                  { return nil }
+
+func (r *registryD) HealthFunc(runtime.Runtime) health.Check {
```
Comment (DmitriyMV) [inline internal/app/machined/pkg/system/services/registryd.go:37]: @smira healthcheck ^^^

**Pull request merged**